File: Dynamic Programming/Java/Minimum Total Distance Traveled.java
// Problem link: https://leetcode.com/problems/minimum-total-distance-traveled/description

// Approach: DP (memo)
// Time: O(n ^ 2 * m)
// Space: O(n * m)
class Solution {
    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {
        Collections.sort(robot);
        Arrays.sort(factory, (a, b) -> Integer.compare(a[0], b[0]));
        List<Integer> factories = new ArrayList<>();

        for(int[] pair : factory) {
            for(int limit = 0; limit < pair[1]; limit++) {
                factories.add(pair[0]);
            }
        }

        int rows = robot.size(), cols = factories.size();
        long[][] DP = new long[rows][cols];

        for(long[] row : DP) {
            Arrays.fill(row, -1);
        }

        return this.dfs(robot, factories, 0, 0, DP);
    }

    private long dfs(List<Integer> robot, List<Integer> factories, int rIdx, int fIdx, long[][] DP) {
        if(rIdx == robot.size()) {
            return 0;
        }
        else if(fIdx == factories.size()) {
            return (long)1e12;
        }
        else if(DP[rIdx][fIdx] != -1) {
            return DP[rIdx][fIdx];
        }

        Long include = Math.abs(robot.get(rIdx) - factories.get(fIdx)) + this.dfs(robot, factories, rIdx + 1, fIdx + 1, DP);
        Long exclude = this.dfs(robot, factories, rIdx, fIdx + 1, DP);

        DP[rIdx][fIdx] = Math.min(include, exclude);
        return DP[rIdx][fIdx];
    }
}

// Approach: DP Tabulation
// Time: O(n ^ 2 * m)
// Space: O(n * m)
class Solution {
    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {
        Collections.sort(robot);
        Arrays.sort(factory, (a, b) -> Integer.compare(a[0], b[0]));
        List<Integer> factories = new ArrayList<>();

        for(int[] pair : factory) {
            for(int limit = 0; limit < pair[1]; limit++) {
                factories.add(pair[0]);
            }
        }

        int robotCount = robot.size(), factoryCount = factories.size();
        long include, exclude;
        long[][] DP = new long[robotCount + 1][factoryCount + 1];

        for(int row = 0; row < robotCount; row++) {
            DP[row][factoryCount] = (long)1e12;
        }

        for(int row = robotCount - 1; row >= 0; row--) {
            for(int col = factoryCount - 1; col >= 0; col--) {
                include = Math.abs(robot.get(row) - factories.get(col)) + DP[row + 1][col + 1];
                exclude = DP[row][col + 1];

                DP[row][col] = Math.min(include, exclude);
            }
        }

        return DP[0][0];
    }
}

// Approach: DP (Space optimized)
// Time: O(n ^ 2 * m)
// Space: O(m)
class Solution {
    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {
        Collections.sort(robot);
        Arrays.sort(factory, (a, b) -> Integer.compare(a[0], b[0]));
        List<Integer> factories = new ArrayList<>();

        for(int[] pair : factory) {
            for(int limit = 0; limit < pair[1]; limit++) {
                factories.add(pair[0]);
            }
        }

        int robotCount = robot.size(), factoryCount = factories.size();
        long include, exclude;
        long[] curr = new long[factoryCount + 1], prev = new long[factoryCount + 1];

        for(int row = robotCount - 1; row >= 0; row--) {
            curr[factoryCount] = (long)1e12;
            for(int col = factoryCount - 1; col >= 0; col--) {
                include = Math.abs(robot.get(row) - factories.get(col)) + prev[col + 1];
                exclude = curr[col + 1];

                curr[col] = Math.min(include, exclude);
            }

            for(int col = 0; col < factoryCount; col++) {
                prev[col] = curr[col];
                curr[col] = 0;
            }

            prev[factoryCount] = (long)1e12;
        }

        return prev[0];
    }
}

File: Dynamic Programming/Python/Minimum Total Distance Traveled.py
# Problem link: https://leetcode.com/problems/minimum-total-distance-traveled/description

# Approach: DP (Memo)
# Time: O(n ^ 2 * m)
# Space: O(n * m)
class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        robot.sort()
        factory.sort()
        factories = []

        for fact in factory:
            factories.extend([fact[0]] * fact[1])
        
        memo = [[-1] * len(factories) for _ in range(len(robot))]
        
        return self.dfs(robot, factories, 0, 0, memo)
        

    def dfs(self, robot, factories, factIdx, robIdx, memo):
        if robIdx == len(robot):
            return 0
        elif factIdx == len(factories):
            return float('inf')
        elif memo[robIdx][factIdx] != -1:
            return memo[robIdx][factIdx]
        
        include = abs(factories[factIdx] - robot[robIdx]) + self.dfs(robot, factories, factIdx + 1, robIdx + 1, memo)
        exclude = self.dfs(robot, factories, factIdx + 1, robIdx, memo)

        memo[robIdx][factIdx] = min(include, exclude)
        return min(include, exclude)

# Approach: DP Tabulation
# Time: O(n ^ 2 * m)
# Space: O(n * m)
class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        robot.sort()
        factory.sort()

        factories = []

        for fact in factory:
            factories.extend([fact[0]] * fact[1])
        
        robotCount, factoryCount = len(robot), len(factories)
        DP = [[0] * (factoryCount + 1) for _ in range(robotCount + 1)]

        for row in range(robotCount):
            DP[row][factoryCount] = float('inf')

        for row in range(robotCount - 1, -1, -1):
            for col in range(factoryCount - 1, -1, -1):
                include = abs(robot[row] - factories[col]) + DP[row + 1][col + 1]
                exclude = DP[row][col + 1]

                DP[row][col] = min(include, exclude)
        
        return DP[0][0]

# Approach: DP Space optimized
# Time: O(n ^ 2 * m)
# Space: O(m)
class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        robot.sort()
        factory.sort()

        factories = []

        for fact in factory:
            factories.extend([fact[0]] * fact[1])
        
        robotCount, factoryCount = len(robot), len(factories)
        prev, curr = [0] * (factoryCount + 1), [0] * (factoryCount + 1)
        curr[factoryCount] = float('inf')

        for row in range(robotCount - 1, -1, -1):
            
            for col in range(factoryCount - 1, -1, -1):
                include = abs(robot[row] - factories[col]) + prev[col + 1]
                exclude = curr[col + 1]

                curr[col] = min(include, exclude)
            
            for col in range(factoryCount):
                prev[col], curr[col] = curr[col], 0
                
            prev[factoryCount] = float('inf')
        
        return prev[0]
